craft_parts/plugins/rust_plugin.py

```diff
@@ -111,32 +118,84 @@ def get_build_snaps(self) -> Set[str]:
     @override
     def get_build_packages(self) -> Set[str]:
         """Return a set of required packages to install in the build environment."""
-        return {"gcc", "git"}
+        return {"curl", "gcc", "git", "pkg-config", "findutils"}
+
+    def _check_rustup(self) -> bool:
+        try:
+            rustup_version = subprocess.check_output(["rustup", "--version"])
+            return "rustup" in rustup_version.decode("utf-8")
+        except (subprocess.CalledProcessError, FileNotFoundError):
+            return False
+
+    def _get_setup_rustup(self, channel: str) -> List[str]:
```

Comment:
> I'll defer to @cmatsuoka and @mr-cal on this as the maintainer and author, but we may want to default to getting rust from the system package manager.

That's a bad idea. Our in-archive Rust toolchain is usually lagging, and Rust is a very fast-moving language that releases a new compiler version every six weeks.



> To keep backwards compatibility, I would guess the logic could be something like:
> 
>     1. If a `rust-deps` part dependency is defined, no-op (maybe error if a `rust-channel` is defined).
> 
>     2. If a `rust-channel` is defined, use rustup.
> 
>     3. If neither of those, include `rustc` and `cargo` in the build packages.

I chose the current behavior because if you do a quick search on the Snapcraft forum, you will find most users are confused by this (they are migrating from `core18`, which they would expect the plugin to install a `rustup` binary.

Reply:
I would like to allow user to choice use system rust or install from rustup with/without pinned rust versions.

Reply:
I like the use of rustup (and have dealt with [exactly the lagging rust versions issue](https://github.com/canonical/charmcraft/blob/main/snap/snapcraft.yaml#L74-L94) in charmcraft — though there `focal-updates` getting a new enough rust ended up resolving it). I've also dealt with the confusion of my core22 snap not having rust even when I used the plugin, so I'm all on board with that!

The backwards compatibility issue is within a single `coreX` base rather than between bases. While I personally prefer the core18 behaviour, we can't break existing `core22` snaps (or charms/rocks that use the plugin).

The two major ways people currently use to get rust on core22 are to install it in a `rust-deps` part or to add `cargo` to `build-packages` (as I did with charmcraft, and also like [sccache](https://github.com/mozilla/sccache/blob/main/snap/snapcraft.yaml#L40) and [alacritty](https://github.com/snapcrafters/alacritty/blob/1f478744ea69f1095c5399e65db2888b9d598ebe/snap/snapcraft.yaml#L75), and we need to ensure that neither of these break existing projects.

Perhaps the following behaviour will be more palatable:

1. If a `rust-deps` part dependency is defined, assume that installs rust and error if there's rustup configuration here.
  a) if there is a `rust-deps` part dependency, maybe log the fact that it's overriding the installation so users have a friendly way to know that they need to rename it
2. If `rust-channel` is undefined, check for `rustc` and `cargo` during the pull step. If they exist **and** `rustup` does not, no-op. (This should work since `build-packages` are installed before the pull step).
3. If we don't have `cargo` or `rustc`, do the `rustup` download/update logic.